### PR TITLE
Stabilize `extended_varargs_abi_support`

### PR DESCRIPTION
**Pull request by Soveu**
I think that is everything? If there is any documentation regarding `extern` and/or varargs to correct, let me know, some quick greps suggest that there might be none.

Tracking issue: https://github.com/rust-lang/rust/issues/100189

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -435,9 +435,6 @@ declare_features! (
     (active, exhaustive_patterns, "1.13.0", Some(51085), None),
     /// Allows explicit tail calls via `become` expression.
     (incomplete, explicit_tail_calls, "1.72.0", Some(112788), None),
-    /// Allows using `efiapi`, `sysv64` and `win64` as calling convention
-    /// for functions with varargs.
-    (active, extended_varargs_abi_support, "1.65.0", Some(100189), None),
```
Comment (Nilstrieb) [inline compiler/rustc_feature/src/active.rs:440]: we don't delete features, we instead move them to `accepted.rs`

Comment (Soveu) [reply compiler/rustc_feature/src/active.rs:440]: Oh, right, I missed the part where I paste this back, thanks

File: compiler/rustc_feature/src/accepted.rs [hunk 2]
```diff
@@ -160,6 +160,9 @@ declare_features! (
     (accepted, explicit_generic_args_with_impl_trait, "1.63.0", Some(83701), None),
     /// Allows arbitrary expressions in key-value attributes at parse time.
     (accepted, extended_key_value_attributes, "1.54.0", Some(78835), None),
+    /// Allows using `efiapi`, `aapcs`, `sysv64` and `win64` as calling
+    /// convention for functions with varargs.
+    (accepted, extended_varargs_abi_support, "1.65.0", Some(100189), None),
```
Comment (kadiwa4) [inline compiler/rustc_feature/src/accepted.rs:165]: use `CURRENT_RUSTC_VERSION` instead of a version number here

**Comment by tgross35** [comment]
Tracking issue: https://github.com/rust-lang/rust/issues/100189

**Comment by Soveu** [comment]
@wesleywiser

**Comment by apiraino** [comment]
r? compiler

File: compiler/rustc_feature/src/accepted.rs [hunk 3]
```diff
@@ -159,6 +159,9 @@ declare_features! (
     (accepted, explicit_generic_args_with_impl_trait, "1.63.0", Some(83701), None),
     /// Allows arbitrary expressions in key-value attributes at parse time.
     (accepted, extended_key_value_attributes, "1.54.0", Some(78835), None),
+    /// Allows using `efiapi`, `aapcs`, `sysv64` and `win64` as calling
+    /// convention for functions with varargs.
+    (accepted, extended_varargs_abi_support, "1.74.0", Some(100189), None),
```
Comment (Skgland) [inline compiler/rustc_feature/src/accepted.rs:164]: As noted in https://github.com/rust-lang/rust/pull/116161#discussion_r1336437170
```suggestion
    (accepted, extended_varargs_abi_support, "CURRENT_RUSTC_VERSION", Some(100189), None),
```

**Comment by Nadrieril** [comment]
I'm on vacation for a few days :)
r? compiler

**Comment by cjgillot** [comment]
Not completely sure which team should be nominated for stabilization, so labeling both in doubt.

**Comment by pnkfelix** [comment]
Just to add on to @cjgillot 's comment above: @wesleywiser and I could not remember earlier today whether T-lang *wants* to own FCP'ing changes like this that are restricted to extending the set of calling-conventions (i.e. the `conv` in `extern "conv" fn foo(...)`), which is largely a detail about what platforms one is interoperating with, and not about changing the expressiveness of the Rust language as a whole in the abstract.

**Comment by apiraino** [comment]
Removing the T-compiler nomination until after T-lang discussed this. Feel free to re-add T-compiler!

@rustbot label -I-compiler-nominated

**Comment by Soveu** [comment]
@cjgillot is there anything else that is blocking us from merging this?

**Comment by cjgillot** [comment]
@Soveu this is waiting for lang team discussion. Once we get their approval (or their free pass to t-compiler to do whatever they want) it will be merged.

**Comment by tmandry** [comment]
Lang team reviewed in our meeting and we're happy to move forward with stabilizing.

Question: Does this include `-unwind` variants for each of the ABIs or not?

@rustbot label -I-lang-nominated -I-lang-easy-decision

**Comment by Soveu** [comment]
> Question: Does this include `-unwind` variants for each of the ABIs or not?

Well, yes. I haven't thought about it initially (as I was mostly focused on `efiapi`, non-unwind `win64` and `sysv64`), but currently `Abi::supports_varargs` looks like this:
```rust
    pub fn supports_varargs(self) -> bool {
        // * C and Cdecl obviously support varargs.
        // * C can be based on Aapcs, SysV64 or Win64, so they must support varargs.
        // * EfiApi is based on Win64 or C, so it also supports it.
        // * System falls back to C for functions with varargs.
        //
        // * Stdcall does not, because it would be impossible for the callee to clean
        //   up the arguments. (callee doesn't know how many arguments are there)
        // * Same for Fastcall, Vectorcall and Thiscall.
        // * Other calling conventions are related to hardware or the compiler itself.
        match self {
            Self::C { .. }
            | Self::Cdecl { .. }
            | Self::System { .. }
            | Self::Aapcs { .. }
            | Self::Win64 { .. }
            | Self::SysV64 { .. }
            | Self::EfiApi => true,
            _ => false,
        }
    }
```
So whether there is `Abi::Win64 { unwind: true }` or `Abi::Win64 { unwind: false }`, it should treat them the same.

Should I rebase the change or will there be some exceptions for `-unwind` calling conventions?

**Comment by Soveu** [comment]
> Removing the T-compiler nomination until after T-lang discussed this. Feel free to re-add T-compiler!
> 
> @rustbot label -I-compiler-nominated

@rustbot label +I-compiler-nominated

**Comment by Soveu** [comment]
lgtm :)
@rustbot label +S-waiting-on-review

**Comment by Soveu** [comment]
uhh, I did not notice that I changed some submodules

**Comment by cjgillot** [comment]
r=me after rebase

**Comment by Soveu** [comment]
@cjgillot the CI checks have passed

**Comment by Soveu** [comment]
@cjgillot I have rebased the change because it had one conflict

**Comment by jieyouxu** [comment]
bors has a quirk where *old* PRs might need to get reopened as bors may have forgotten about the PR.

**Comment by matthiaskrgr** [comment]
@bors r- 
seems to have failed in https://github.com/rust-lang/rust/pull/133542#issuecomment-2504720786

**Comment by Soveu** [comment]
> @bors r- seems to have failed in [#133542 (comment)](https://github.com/rust-lang/rust/pull/133542#issuecomment-2504720786)

Yeah, it is this change... already pushed a corrected version

**Comment by ehuss** [comment]
Can you please post a PR to update the documentation? I assume it would go in https://github.com/rust-lang/reference/blob/master/src/types/function-pointer.md.

**Comment by traviscross** [comment]
@bors r-

We shouldn't merge this until there is a Reference PR.

**Comment by traviscross** [comment]
@rustbot labels -S-blocked
@bors r=cjgillot

There's an approved Reference PR now (https://github.com/rust-lang/reference/pull/1687#pullrequestreview-2470609270), so this is good to go.

**Comment by chorman0773** [comment]
This is in beta but not yet stable. It should probably be reverted as <https://godbolt.org/z/fYzxa8xW5> compiles with this but probably ought not to, at least on targets like i686-pc-windows-msvc where the underlying abi of `extern "system"` is incompatible with varargs.

**Comment by chorman0773** [comment]
@rustbot label +I-compiler-nominated +I-lang-nominated

Should this be reverted (and backported) to allow addressing whether `extern "system"` should allow varargs functions, and when?

**Comment by workingjubilee** [comment]
moving that discussion to a new issue.

**Pull request merged**